File: Decorator Pattern/decorator.ts
interface IToolbox {
  weight: number;
  toolCount: number;
}

class Toolbox implements IToolbox {
  private _weight = 200;
  private _toolCount = 0;

  get weight() {
    return this._weight;
  }

  get toolCount() {
    return this._toolCount;
  }
}

class Wrenches implements IToolbox {
  private _toolbox: IToolbox;
  private _weight = 3.5;
  private _toolCount = 7;

  constructor(toolbox: IToolbox) {
    this._toolbox = toolbox;
  }

  get weight() {
    return this._weight + this._toolbox.weight;
  }

  get toolCount() {
    return this._toolCount + this._toolbox.toolCount;
  }
}

class Sockets implements IToolbox {
  private _toolbox: IToolbox;
  private _weight = 1.8;
  private _toolCount = 15;

  constructor(toolbox: IToolbox) {
    this._toolbox = toolbox;
  }

  get weight() {
    return this._weight + this._toolbox.weight;
  }

  get toolCount() {
    return this._toolCount + this._toolbox.toolCount;
  }
}

class Screwdrivers implements IToolbox {
  private _toolbox: IToolbox;
  private _weight = 2.5;
  private _toolCount = 10;

  constructor(toolbox: IToolbox) {
    this._toolbox = toolbox;
  }

  get weight() {
    return this._weight + this._toolbox.weight;
  }

  get toolCount() {
    return this._toolCount + this._toolbox.toolCount;
  }
}

const toolbox = new Screwdrivers(new Sockets(new Wrenches(new Toolbox())));

console.log(toolbox.weight); // 207.8
console.log(toolbox.toolCount); // 32

File: Decorator Pattern/decorator.rb
# frozen_string_literal: true

class Toolbox
  WEIGHT = 200

  def tool_count
    0
  end

  def weight
    WEIGHT
  end
end

class Wrenches
  TOOL_COUNT = 7
  WEIGHT = 3.5

  attr_reader :toolbox

  def initialize(toolbox)
    @toolbox = toolbox
  end

  def tool_count
    TOOL_COUNT + toolbox.tool_count
  end

  def weight
    WEIGHT + toolbox.weight
  end
end

class Sockets
  TOOL_COUNT = 15
  WEIGHT = 1.8

  attr_reader :toolbox

  def initialize(toolbox)
    @toolbox = toolbox
  end

  def tool_count
    TOOL_COUNT + toolbox.tool_count
  end

  def weight
    WEIGHT + toolbox .weight
  end
end

class Screwdrivers
  TOOL_COUNT = 10
  WEIGHT = 2.5

  attr_reader :toolbox

  def initialize(toolbox)
    @toolbox = toolbox
  end

  def tool_count
    TOOL_COUNT + toolbox.tool_count
  end

  def weight
    WEIGHT + toolbox.weight
  end
end

toolbox = Screwdrivers.new(
  Sockets.new(
    Wrenches.new(
      Toolbox.new
    )
  )
)

p toolbox.tool_count
p toolbox.weight

# >> 32
# >> 207.8
